man/fi_profile.3.md

```diff
@@ -0,0 +1,281 @@
+  The information referenced by the param is event specific.
+
+*size*
+: The amount of the data in the "param" when the callback is called.
+
+*data*
+: The current value of a variable.
+
+# DESCRIPTION
+
+Fabric profiling interface allows users to obtain configuration or runtime
+properties from the providers in the form of "variables". Providers 
+define variables and make them readable to users through the profiling 
+interface. Providers define events and use them to report when an operation 
+has occured or a state changed where one or more variables may change their 
+values. Users can register callback functions to one or more events, and 
+receive notifications from providers once an event occurs.
+
+In order to support the profiling interface, an endpoing or a domain in a 
+provider needs to maintain a fid_profile object, and implement the profiling 
+operations there. To start a profiling, an open call must be made on the 
+target endpoint or domain object to obtain the fid_profile object.  If the 
+target object does not support profiling, the open call will fail with error 
+code -FI_ENOSYS. Once a fid_profile object is obtained successfully, it can be 
+used to query available variables or events, and read variables' values.
+
+# VARIABLES
+
+The profiling variables are defined in providers and are accessible through 
+the profiling interface. Variables are specified using struct 
+fi_profile_desc described below. A variable can be a configuration setting, 
+a runtime state or a counter associated with an endpoint or a domain. 
+Some variables are common and available in most providers, while others are 
+provider-specific, and are only available in one provider. Users can query 
+available variables and their values using the profiling interface.
+
+Provider specific variables are outside the scope of this man page. libfabric 
+defined variables and their datatypes are described below.
+
+## FI_VAR_UNEXP_MSG_CNT (uint64_t)
+
+This variable returns the current number of unexpected messages that have been
+received by the provider which are waiting for matching receive buffers.
+
+# EVENTS
+
+Profiling events notify users that an operation has occurred or the state of
+an object has changed. Events typically indicate that the values of one or 
+more variables have been modified. Similar to variables, events may be 
+pre-defined by libfabric or be provider specific. Users can register a 
+callback to be notified when an event occurs.  libfabric defined events 
+and the data that corresponds with the event, if any, are described below.
+
+Applications may read variable values from the event callback, but may not 
+register or unregister for events, query for profiling variable or events, 
+or call the profile lock and unlock functions
+
+libfarbic pre-defined events include:
+
+## FI_EVENT_UNEXP_MSG_RECVD 
+
+This event indicates that an unexpected message was received and has been
+queued by the provider. 
```

Comment:
I was initially thinking of making the event simple with no param, completely depends on the user to pull the data. 

If defining this event data struct,  we need to add it into fi_profile.h. 
Can we expect for the same event, any provider will notify with the same type of "param"? If this is the case, we can define an event as <event, data_type, data_size>,  different events can be extended from a base "data_type" for easy expansion.

Reply:
Any structure that we define here will need to be added to the header file.  And we need to define that the same parameter be used by all providers for common events.  Otherwise, there's nothing an app can realistically do with any of this.  It's the same as reading a variable.